plugins/inputs/win_wmi/win_wmi.go

```diff
@@ -0,0 +1,277 @@
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/filter"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+const sampleConfig = `
+  ## By default, this plugin returns no results.
+  ## Uncomment the example below or write your own as you see fit.
+  ## [[inputs.win_wmi]]
+  ##   name_prefix = "win_wmi_"
+  ##   [[inputs.win_wmi.query]]
+  ##     Namespace = "root\\cimv2"
+  ##     ClassName = "Win32_Volume"
+  ##     Properties = ["Name", "Capacity", "FreeSpace"]
+  ##     Filter = 'NOT Name LIKE "\\\\?\\%"'
+  ##     TagPropertiesInclude = ["Name"]
+`
+
+// Query struct
+type Query struct {
+	Query                string   `toml:"query"`
+	Namespace            string   `toml:"Namespace"`
+	ClassName            string   `toml:"ClassName"`
+	Properties           []string `toml:"Properties"`
+	Filter               string   `toml:"Filter"`
+	TagPropertiesInclude []string `toml:"TagPropertiesInclude"`
+	tagFilter            filter.Filter
+}
+
+var lock sync.Mutex
+
+// Wmi struct
+type Wmi struct {
+	Queries []Query `toml:"query"`
+}
+
+// S_FALSE is returned by CoInitializeEx if it was already called on this thread.
+const sFalse = 0x00000001
+
+func oleInt64(item *ole.IDispatch, prop string) (int64, error) {
+	v, err := oleutil.GetProperty(item, prop)
+	if err != nil {
+		return 0, err
+	}
+	defer func() { _ = v.Clear() }()
+
+	i := v.Val
+	return i, nil
+}
+
+// Description function
+func (s *Wmi) Description() string {
+	return "returns WMI query results as metrics"
+}
+
+// SampleConfig function
+func (s *Wmi) SampleConfig() string {
+	return sampleConfig
+}
+
+// build a WMI query
+func BuildWmiQuery(q Query) (string, error) {
+	var wql string
+	var b bytes.Buffer
+	_, err := b.WriteString("SELECT ")
+	if err != nil {
+		return wql, err
+	}
+
+	_, err = b.WriteString(strings.Join(q.Properties, ", "))
+	if err != nil {
+		return wql, err
+	}
+
+	_, err = b.WriteString(" FROM ")
+	if err != nil {
+		return wql, err
+	}
+
+	_, err = b.WriteString(q.ClassName)
+	if err != nil {
+		return wql, err
+	}
+
+	if len(q.Filter) > 0 {
+		_, err = b.WriteString(" WHERE ")
+		if err != nil {
+			return wql, err
+		}
+
+		_, err = b.WriteString(q.Filter)
+		if err != nil {
+			return wql, err
+		}
+	}
+	wql = b.String()
+
+	return wql, nil
+}
+
+func DoQuery(q Query, acc telegraf.Accumulator) error {
+	lock.Lock()
+	defer lock.Unlock()
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	tags := map[string]string{}
+	fields := map[string]interface{}{}
+
+	// init COM
+	err := ole.CoInitializeEx(0, ole.COINIT_MULTITHREADED)
+	if err != nil {
+		oleCode := err.(*ole.OleError).Code()
+		if oleCode != ole.S_OK && oleCode != sFalse {
+			return err
+		}
+	}
+	defer ole.CoUninitialize()
+
+	unknown, err := oleutil.CreateObject("WbemScripting.SWbemLocator")
+	if err != nil {
+		return err
+	} else if unknown == nil {
+		panic("CreateObject returned nil")
+	}
```

Comment:
Please don't create panics. Telegraf should be able to keep running other plugins while one has an issue.